ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2020, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Prometheus Metrics Exporter for OpenTelemetry."""
+
+import collections
+import logging
+import re
+from typing import Sequence
+
+from prometheus_client import start_http_server
+from prometheus_client.core import (
+    REGISTRY,
+    CollectorRegistry,
+    CounterMetricFamily,
+    GaugeMetricFamily,
+    UnknownMetricFamily,
+)
+
+from opentelemetry.metrics import Counter, Gauge, Measure, Metric
+from opentelemetry.sdk.metrics.export import (
+    MetricRecord,
+    MetricsExporter,
+    MetricsExportResult,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class PrometheusMetricsExporter(MetricsExporter):
+    """Prometheus metric exporter for OpenTelemetry.
+
+    Args:
+        port: Port number to listen.
+        address: Endpoint address (default is localhost).
+        prefix: single-word application prefix relevant to the domain the metric belongs to.
+    """
+
+    def __init__(self, port: int = 8000, address: str = "", prefix: str = ""):
+        self._port = port
+        self._address = address
+        self._collector = CustomCollector(prefix)
+
+        start_http_server(port=self._port, addr=self._address)
+        REGISTRY.register(self._collector)
+
+    def export(
+        self, metric_records: Sequence[MetricRecord]
+    ) -> MetricsExportResult:
+        self._collector.add_metrics_data(metric_records)
+        return MetricsExportResult.SUCCESS
+
+    def shutdown(self) -> None:
+        REGISTRY.unregister(self._collector)
+
+
+class CustomCollector:
+    """ CustomCollector represents the Prometheus Collector object
+        https://github.com/prometheus/client_python#custom-collectors
+    """
+
+    def __init__(self, prefix: str = ""):
+        self._prefix = prefix
+        self._metrics_to_export = collections.deque()
+        self._non_letters_nor_digits_re = re.compile(
+            r"[^\w]", re.UNICODE | re.IGNORECASE
+        )
+
+    def add_metrics_data(self, metric_records: Sequence[MetricRecord]):
+        self._metrics_to_export.append(metric_records)
+
+    def collect(self):
+        """Collect fetches the metrics from OpenTelemetry
+        and delivers them as Prometheus Metrics.
+        Collect is invoked every time a prometheus.Gatherer is run
```

Comment:
Is there a link to some API that shows how this works?

Reply:
The custom collector will use same logic as regular Prometheus Collector, pulling mechanism is configured by time intervals but is also triggered by other input like the user refreshing Prometheus visualization UI

https://github.com/prometheus/client_python#custom-collectors
https://github.com/prometheus/client_python/blob/master/prometheus_client/registry.py